### PR TITLE
SLING-10754 : Missing entries in filter.xml (test-case illustrating the problem)

### PR DESCRIPTION
**Pull request by anchela**

### CODE REVIEW

**Review by karlpauls — APPROVED** [review]
Looks very good - thanks a lot!

File: src/test/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    private final WorkspaceFilter base = mock(WorkspaceFilter.class);
+
+    private final PathFilterSet pfs = new PathFilterSet(TEST_ROOT_PATH);
+    private final List<PathFilterSet> filterSets = new ArrayList<>();
+    
+    @Before
+    public void before() throws Exception {
+        pfs.setImportMode(ImportMode.UPDATE);
+        pfs.addExclude(new DefaultPathFilter("test/exclude"));
+        pfs.addInclude(new DefaultPathFilter("test/include"));
```
Comment (kwin) [inline src/test/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilderTest.java:59]: this is redundant, if the first filter is an exclude an implicit "include all" is assumed (https://jackrabbit.apache.org/filevault/filter.html#Include_and_Exclude_Elements)

File: src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+     * - it's root path is a sibling of a manually extracted paths hidden in a .content.xml that get 
+     *   installed despite not being covered by the filter (issue in older versions of Jackrabbit FileVault)).
+     * 
+     * @param pfs A {@link PathFilterSet} of the original base filter.
+     * @return {@code true} if the given {@code PathFilterSet} is still relevant for the new {@code WorkspaceFilter} given
+     * the new content of the converted package as reflected by the recorded paths.
+     */
+    private boolean coversConvertedPath(@NotNull PathFilterSet pfs) {
+        return cpPaths.stream().anyMatch(pfs::covers) ||
+                // test if a extracted path matches and workaround sibling-issue in older versions Jackrabbit FileVault
```
[the remark this answers is not in the document]
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java:124]: hi @kwin i asked you for the link in SLING-10760. i searched in filevault jira and didn't find which issue you were referring to that fixed the sibling problem in newer versions of filevault.

File: src/test/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilderTest.java (hunk 1 above)
Comment (anchela) [reply src/test/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilderTest.java:59]: this is unit-test for the builder and not an integration test. it's used to verify that the include/exclude entries are copied over not that they are sensible.

File: src/main/java/org/apache/sling/feature/cpconverter/shared/AbstractJcrNodeParser.java [hunk 3]
```diff
@@ -27,15 +27,14 @@
...
-import org.apache.sling.feature.cpconverter.ConverterException;
 import org.jetbrains.annotations.NotNull;
 import org.xml.sax.Attributes;
 import org.xml.sax.SAXException;
 import org.xml.sax.helpers.DefaultHandler;
 
 public abstract class AbstractJcrNodeParser<O> extends DefaultHandler {
 
-    private static final String JCR_ROOT = "jcr:root";
+    public static final String JCR_ROOT = "jcr:root";
```
[the remark this answers is not in the document]
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/shared/AbstractJcrNodeParser.java:37]: hi @kwin , which constant you are referring to? i checked org.apache.jackrabbit.vault.util.Constants.java and it only contains `String ROOT_DIR = "jcr_root"`. i would have used a "jcr:root" constant if i had found one :)

File: src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java:124]: I am not aware of any sibling problem. If that still is a problem in the latest FileVault, please report it!

File: src/main/java/org/apache/sling/feature/cpconverter/shared/AbstractJcrNodeParser.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/feature/cpconverter/shared/AbstractJcrNodeParser.java:37]: Indeed, there isn't. I am gonna add it to https://jackrabbit.apache.org/filevault/apidocs/org/apache/jackrabbit/vault/util/Constants.html

File: src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java:124]: hi @kwin in SLING-10760 you stated:
```
With which version of FileVault have you tested importing that package (AEM uses a pretty outdated one)? Everything not covered by the filter.xml should not be installed! The only edge case is nodes represented by empty elements (http://jackrabbit.apache.org/filevault/docview.html#Empty_Elements) in enhanced docview which will not be removed, but this is not really related to the filter.xml question.

There is currently a bug related to authorization nodes (JCRVLT-522) which are applied independent if they are contained in the filter rules. Not sure we can ever fix this, as a lot of packages probably now rely on this bug!
```
my reading was that importing non-covered nodes as i see it in the test is therefore an issue that has been fixed in Jackrabbit FileVault. I searched JIRA but was not able to find the corresponding ticket. Therefore I asked you for the corresponding link in https://issues.apache.org/jira/browse/SLING-10760?focusedCommentId=17405809&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17405809.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+     * 
+     * @param repositoryPath The base repository path used to build absolute paths from the parsed elements.
+     */
+    public DefaultEntryParser(@NotNull String repositoryPath) {
+        currentPath.push(repositoryPath);
+    }
+
+    @Override
+    public void startElement(String uri, String localName, String qName, Attributes attributes) {
+        if (JCR_ROOT.equals(qName)) {
```
Comment (kwin) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java:48]: The reason why there is no constant is that the namespace mapping is allowed to be different in enhanced docview files (https://jackrabbit.apache.org/filevault/docview.html#Root_Element). Although unlikely the root element may also be named "my:root" as long as "my" is a prefix for namespace url "http://www.jcp.org/jcr/1.0". You should really work with fully qualified named here.

File: src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java:124]: I though you tested specifically with `rep:policy` nodes, which have some inconsistencies during import as outlined in JCRVLT-522. Are regular content nodes also installed if they are outside the filter? Would be good to have a test case in FileVault in that case. I am not aware of such issues (even in older versions of FileVault).

[the remark this answers is not in the document]
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java:124]: @kwin feel free to use the packages as created for this ticket for tests in jackrabbit filevault. i will adjust the comment accordingly.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java (hunk 4 above)
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java:48]: the sling feature-cp-converter ignores namespace remapping all over the place. please file a separate ticket to have it fixed everywhere.

File: src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/feature/cpconverter/vltpkg/WorkspaceFilterBuilder.java]: Sorry I won't. If you cannot even afford the time to do a proper bug report for Jackrabbit FileVault it just won't be fixed. Why do you guys care so little about JR FileVault?

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java (hunk 4 above)
Comment (karlpauls) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java:48]: I think I agree, we should investigate that in a different issue.

Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/DefaultEntryParser.java:48]: @karlpauls , i created SLING-10768 to keep track of this separate issue.

**Pull request merged**